### PR TITLE
Feature: reload pinned images

### LINKED ISSUE #167: Add mechanism to automatically reload slice images to camera module
It would be nice if we can reload the camera module's pinned images from disk after restarting acq4. Possible mechanisms:
- "pin selected images" button in camera module that loads files selected in dta manager module
- "reload previous state" button in manager that automatically reloads previous state on shutdown (for example, pined images, loaded modules, window positions, etc.). Modules could provide saveState / loadState methods to support this.
- "restart acq4" button in manager module that does an automatic save, quit, restart, reload

### PR DESCRIPTION
**Pull request by outofculture**
closes #167

### CODE REVIEW

File: acq4/filetypes/FileType.py [hunk 1]
```diff
@@ -29,32 +26,35 @@ def write(cls, data, dirHandle, fileName, **args):
...
         """Return priority value if the file can be read by this class.
-        Otherwise return False.
+        Otherwise, return False.
         The default implementation just checks for the correct name extensions."""
         name = fileHandle.name()
-        for ext in cls.extensions:
-            if name[-len(ext):].lower() == ext.lower():
-                return cls.priority
-        return False
+        return next(
```
[the remark this answers is not in the document]
Comment (outofculture) [reply acq4/filetypes/FileType.py:42]: oh, yeah, same thing. `next` is only mildly my preference, so I'm cool reverting this.

File: acq4/modules/DataManager/DataManagerModule.py [hunk 2]
```diff
@@ -112,14 +106,22 @@ def setLogDir(self):
             d = d.parent()
         self.manager.setLogDir(d)
 
+    def loadPinnedImages(self):
+        cam_mod = self.manager.getModule("Camera")
+        current_dir = self.manager.getCurrentDir()
+        for f in current_dir.ls():
+            if f.endswith('.tif'):
+                f = current_dir[f]
+                frame = Frame(f.read(), f.info().deepcopy())
```
Comment (campagnola) [inline acq4/modules/DataManager/DataManagerModule.py:115]: Frame.load(fileHandle) automatically checks for linked files?

File: acq4/util/DataManager.py [hunk 3]
```diff
@@ -550,35 +534,35 @@ def mkdir(self, name, autoIncrement=False, info=None):
...
         
-    def getDir(self, subdir, create=False, autoIncrement=False):
-        """Return a DirHandle for the specified subdirectory. If the subdir does not exist, it will be created only if create==True"""
+    def getDir(self, subdir, create=False, autoIncrement=False) -> "DirHandle":
+        """Return a DirHandle for the specified subdirectory. If the subdir does not exist, it will be created only 
+        if create==True"""
         with self.lock:
             ndir = os.path.join(self.path, subdir)
-            if not create or os.path.isdir(ndir):
+            if os.path.isdir(ndir):
```
Comment (campagnola) [inline acq4/util/DataManager.py:560]: `if not create or ...`

File: acq4/util/imaging/contrast_ctrl.py [hunk 4]
```diff
@@ -52,25 +51,33 @@ def zoomToImage(self):
...
-        self.alpha = val / self.ui.alphaSlider.maximum()  # slider only works in integers and we need a 0 to 1 value
+        self.alpha = val / self.ui.alphaSlider.maximum()  # slider only works in integers, and we need a 0 to 1 value
         self.imageItem.setOpacity(self.alpha)
 
     def getLevels(self):
         return self.ui.histogram.getLevels()
 
+    def save(self, img, dh: DirHandle) -> str:
+        if self._cachedSaveName is None:
+            lut = self.ui.histogram.getLookupTable(img)
```
Comment (campagnola) [inline acq4/util/imaging/contrast_ctrl.py:75]: Instead just use `self.ui.histogram.saveState()` and this should be small enough to go in .index directly

File: acq4/util/imaging/frame.py [hunk 5]
```diff
@@ -65,15 +67,44 @@ def mapFromFrameToGlobal(self, obj):
         """
         return self.globalTransform().map(obj)
     
-    def saveImage(self, dh, filename):
+    def saveImage(self, dh, filename, backgroundControl=None, contrastControl=None):
```
Comment (campagnola) [inline acq4/util/imaging/frame.py:70]: Pass in metadata here rather than bg/contrast controls.

File: acq4/util/imaging/imaging_ctrl.py [hunk 6]
```diff
@@ -151,8 +150,8 @@ def newFrame(self, frame):
 
     def saveFrameClicked(self):
         if self.ui.linkSavePinBtn.isChecked():
-            self.addPinnedFrame()
-        self.recordThread.saveFrame()
```
Comment (campagnola) [inline acq4/util/imaging/imaging_ctrl.py:155]: Get all metadata fro controls at this point to make sure values are current and we don't try to access widgets from another thread

File: acq4/util/imaging/record_thread.py [hunk 7]
```diff
@@ -75,11 +73,18 @@ def recording(self):
         """
         return self._recording
 
-    def saveFrame(self):
-        """Ask the recording thread to save the most recently acquired frame.
-        """
+    def saveFrame(self, bgCtrl, contrastCtrl):
```
Comment (campagnola) [inline acq4/util/imaging/record_thread.py:76]: Get metadata from bg/contrast controls at this point; don't let background threads access widgets

File: acq4/modules/DataManager/DataManagerModule.py (hunk 2 above)
Comment (outofculture) [reply acq4/modules/DataManager/DataManagerModule.py:115]: `Frame.load` isn't a thing that exists... Maybe you are suggesting I make that? `Frame` isn't tied to the `FileHandle.read` architecture, but it totally could be, if you want.

File: acq4/util/imaging/imaging_ctrl.py [hunk 8]
```diff
@@ -258,18 +263,21 @@ def addPinnedFrame(self):
...
         else:
             z = self.pinnedFrames[-1].zValue() + 1
         im.setZValue(z)
-
+        im.sigRemoveRequested.connect(self.removePinnedFrame)
         self.pinnedFrames.append(im)
         view = self.frameDisplay.imageItem().getViewBox()
         if view is not None:
             view.addItem(im)
-        im.setTransform(self.frameDisplay.currentFrame.globalTransform().as2D())
```
Comment (outofculture) [inline acq4/util/imaging/imaging_ctrl.py:272]: Does this need to be handled somewhere else? Did I already handle this?

File: acq4/devices/Pipette/pipette.py [hunk 9]
```diff
@@ -494,55 +492,6 @@ def getRecordingChambers(self) -> List[RecordingChamber]:
         man = getManager()
         return [man.getDevice(d) for d in self.config.get('recordingChambers', [])]
 
-    def startRecording(self):
-        """Return an object that records all motion updates from this pipette
-        """
-        return PipetteRecorder(self)
-
-
-class PipetteRecorder:
```
Comment (campagnola) [inline acq4/devices/Pipette/pipette.py:503]: I want to liiiiive

**Pull request merged**